### PR TITLE
Allow explicit set of method on Dataset#save

### PR DESCRIPTION
**Pull request by ryanseys**
Should satisfy the needs of #384

### CODE REVIEW

**Comment by stephenplusplus** [comment]
The way you implemented this is quite brilliant in its simplicity!

Do you think we should branch out the API to add these methods in? `dataset.insert/upsert/etc` that would just wrap the longer `method: "insert/upsert/etc"` call?

**Comment by ryanseys** [comment]
I wasn't sure what was preferred here, so your suggestions are very welcome! This is one less-obvious technique, however if you think more value could come from separate methods, let's do that instead.

**Comment by stephenplusplus** [comment]
My only thought is `save` is magic, so `insert/upsert/etc` is declarative. I just think it would be expected that we have those methods. Your implementation is still perfect, we can just add the convenience methods (which we're growing quite a collection of!)

**Comment by ryanseys** [comment]
Should we merge this and then we can open an issue to add the declarative methods?

File: test/datastore/request.js [hunk 1]
```diff
@@ -205,6 +205,25 @@ describe('Request', function() {
...
+    it('should save with specific method', function(done) {
+      request.makeReq_ = function(method, req, callback) {
+        assert.equal(method, 'commit');
+        assert.equal(req.mutation.insert.length, 1);
+        assert.equal(req.mutation.update.length, 1);
+        assert.equal(req.mutation.insert[0].property[0].name, 'k');
+        assert.equal(
+          req.mutation.insert[0].property[0].value.string_value, 'v');
+        assert.equal(req.mutation.update[0].property[0].name, 'k2');
+          assert.equal(
```
Comment (stephenplusplus) [inline test/datastore/request.js]: Extra indent.

File: lib/datastore/request.js [hunk 2]
```diff
@@ -252,15 +271,35 @@ DatastoreRequest.prototype.save = function(entities, callback) {
 
       ent.key = entity.keyToKeyProto(entityObject.key);
 
-      if (entity.isKeyComplete(entityObject.key)) {
-        acc.upsert.push(ent);
+      if (method) {
+        switch(method) {
```
Comment (stephenplusplus) [inline lib/datastore/request.js]: `switch (method)`

File: regression/datastore.js [hunk 3]
```diff
@@ -119,6 +119,38 @@ describe('datastore', function() {
...
+      var postKey = ds.key('Post');
+
+      ds.save({ key: postKey, data: post }, function(err) {
+        assert.ifError(err);
+
+        // The key's path should now be complete.
+        assert(postKey.path[1]);
+
+        ds.save({ key: postKey, method: 'insert', data: post }, function(err) {
+          assert.notEqual(err, null); // should fail insert
```
Comment (stephenplusplus) [inline regression/datastore.js]: I like!

**Comment by stephenplusplus** [comment]
SGTM. Minor formatting comments, then self-merging is cool with me.

File: lib/datastore/request.js [hunk 4]
```diff
@@ -224,6 +226,21 @@ DatastoreRequest.prototype.get = function(keys, callback) {
...
+ * //-
+ * var userKey = dataset.key(['User', 'chilts']);
+ *
+ * dataset.save([
+ *   {
+ *     key: userKey,
+ *     method: 'insert', // force the method to 'insert'
+ *     data: {
+ *       fullName: 'Andrew Chilton'
+ *     }
```
Comment (chilts) [inline lib/datastore/request.js]: I don't mind if you change this to another (made up) name. :)

Comment (stephenplusplus) [inline lib/datastore/request.js]: No way, you're part of this library forever now. https://github.com/GoogleCloudPlatform/gcloud-node/commit/d3fe788b6a4c59269d17dad246f108d8160a7b3b

Seriously though, I just used your name there to be funny. If you want me to remove the honorable mentions now, I'll be happy to. :+1:

**Pull request merged**